### PR TITLE
Correction/Use a serializable interface Settings instead of Queries implementing Serializable

### PR DESCRIPTION
**Pull request by so7fie**
This PR is because some of the implementations of Query<T> cannot be `Serializable`. We chose then to use classes `BookSettings` and `SaleSettings` as serializable containers for the query parameters of `BookQuery` and `SaleQuery`. 

Now `BookSettings` and `SaleSettings` are used as extraParameter passed from one activity to another instead of `BookQuery` and `SaleQuery`.

Edit 1 (more explanations for the purpose of `Settings`) : 
When one want to pass a parameter from one Activity to another one, the said parameter need to implement `Serializable` interface (https://stackoverflow.com/questions/2736389/how-to-pass-an-object-from-one-activity-to-another-on-android). Until now, all the `Query<T>` interfaces did implement `Serializable`, but this cannot be anymore the case because `Query` now depends on the internal state of the database, and thus we cannot simply serialize a query. To address that, we use the data class `Settings` which is just all the parameters of the query along with their corresponding values, and from now on, this is the parameter we will pass from one Activity to another one. A query for a database can be build given any `Setting` object.

Edit 2:
Joshua changed the `Sale` objects in Firebase. The tests for `SaleDatabase` wont work anymore in any of the branch until the pr of Joshua is merged with main. 

## Pull reviewers stats
Stats for the last 180 days:
|                                                                                                                                                               | User            | Total reviews | Median time to review                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | Total comments |
| ------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------- | ------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------- |
| <a href=https://github.com/Huntsekkern><img src="https://avatars.githubusercontent.com/u/44240415?v=4" width="20"></a>                                        | Huntsekkern     | 11            | [50m](https://app.flowwer.dev/charts/review-time/~(u~(i~'44240415~n~'Huntsekkern)~p~180~r~(~(d~'qqkomy~t~'z8i)~(d~'qqkozj~t~'zl3)~(d~'qqkp4h~t~'zq1)~(d~'qqvo8r~t~'14d)~(d~'qpuo7b~t~'mk)~(d~'qq6uip~t~'3j5)~(d~'qpt0zy~t~'2wmx)~(d~'qpt88n~t~'33vm)~(d~'qptd1d~t~'29r)~(d~'qptd3y~t~'2cc)~(d~'qpumtn~t~'1745)~(d~'qpt4tp~t~'2b1)~(d~'qq8faa~t~'qv)~(d~'qq8fiz~t~'bv)~(d~'qqjij5~t~'i6m)~(d~'qrihqg~t~'1wf3)~(d~'qrm9wa~t~'17x)~(d~'qrma4k~t~'1g7)~(d~'qrmajv~t~'1vi)~(d~'qrmb5j~t~'1cc))))                                                          | **31**         |
| <a href=https://github.com/Snowflake8><img src="https://avatars.githubusercontent.com/u/44499990?v=4" width="20"></a>                                         | Snowflake8      | 12            | [6h 1m](https://app.flowwer.dev/charts/review-time/~(u~(i~'44499990~n~'Snowflake8)~p~180~r~(~(d~'qq6k9y~t~'8hk)~(d~'qq6kh9~t~'8ov)~(d~'qq6r5v~t~'6b)~(d~'qpt3s9~t~'2zf8)~(d~'qqjlvf~t~'b)~(d~'qpx29n~t~'1vgi)~(d~'qq0q7g~t~'5jeb)~(d~'qqdm38~t~'1fge)~(d~'qq6ncv~t~'8609)~(d~'qq8el9~t~'y2h)~(d~'qqjh1g~t~'gox)~(d~'qqjh87~t~'7lg)~(d~'qqsnpq~t~'qn)~(d~'qqxzb3~t~'ikt)~(d~'qrm4bb~t~'1ljt)~(d~'qrkbo5~t~'88t)~(d~'qrm4eh~t~'1rq))))                                                                                                                 | 18             |
| <a href=https://github.com/RaulinN><img src="https://avatars.githubusercontent.com/u/44499915?u=c029696ff2de0902f905fbf35188359e94e08d88&v=4" width="20"></a> | RaulinN         | 11            | [3h 8m](https://app.flowwer.dev/charts/review-time/~(u~(i~'44499915~n~'RaulinN)~p~180~r~(~(d~'qpuwak~t~'8pt)~(d~'qq73am~t~'28)~(d~'qptcqf~t~'1yt)~(d~'qpq5rb~t~'5xjz)~(d~'qqr6wy~t~'aju)~(d~'qq7qbn~t~'1bz)~(d~'qq73ig~t~'jy6)~(d~'qq7vmj~t~'1t5)~(d~'qqhz4w~t~'87bj)~(d~'qq73x3~t~'8mkh)~(d~'qq821x~t~'lj5)~(d~'qreyev~t~'f)~(d~'qqj0lo~t~'u5)~(d~'qqjmtu~t~'n2b)~(d~'qrm5d5~t~'2qe))))                                                                                                                                                             | 12             |
| <a href=https://github.com/so7fie><img src="https://avatars.githubusercontent.com/u/10992808?v=4" width="20"></a>                                             | so7fie          | 11            | [21h 30m](https://app.flowwer.dev/charts/review-time/~(u~(i~'10992808~n~'so7fie)~p~180~r~(~(d~'qqkh75~t~'dr0)~(d~'qqrud7~t~'7qx2)~(d~'qpuy0b~t~'1v)~(d~'qq7qp0~t~'hf)~(d~'qq0t0v~t~'5m7q)~(d~'qq0t2c~t~'5m97)~(d~'qq0p5z~t~'5hby)~(d~'qqew7j~t~'2pkp)~(d~'qqf92s~t~'32fy)~(d~'qqjbho~t~'1ux)~(d~'qqy5l4~t~'5im1)~(d~'qrmbez~t~'1ls)~(d~'qr8kh3~t~'1kfr)~(d~'qrm9jc~t~'1npk)~(d~'qrm6v4~t~'6ai))))                                                                                                                                                    | 11             |
| <a href=https://github.com/DGoedtkindtEPFL><img src="https://avatars.githubusercontent.com/u/61052800?v=4" width="20"></a>                                    | DGoedtkindtEPFL | **13**        | [19h 49m](https://app.flowwer.dev/charts/review-time/~(u~(i~'61052800~n~'DGoedtkindtEPFL)~p~180~r~(~(d~'qqioj1~t~'2x8a)~(d~'qqionj~t~'2xcs)~(d~'qqioou~t~'2xe3)~(d~'qqipcq~t~'2y1z)~(d~'qqru2n~t~'7qmi)~(d~'qqxoq1~t~'21ln)~(d~'qq6rfm~t~'fn8)~(d~'qplma9~t~'1e2x)~(d~'qpuugt~t~'am9h)~(d~'qpuy16~t~'2q)~(d~'qpuoy1~t~'co)~(d~'qpup0b~t~'19at)~(d~'qqjmgc~t~'dy)~(d~'qqrtyv~t~'xlr)~(d~'qq6s0w~t~'8gm)~(d~'qqjjxz~t~'9s4m)~(d~'qq9yj2~t~'1i5c)~(d~'qrm4pz~t~'1iw7)~(d~'qrm4za~t~'1j5i)~(d~'qrm5eq~t~'1jky)~(d~'qrmfv4~t~'1u1c)~(d~'qrmhyq~t~'1cp)))) | 7              |
| <a href=https://github.com/Laniakea1999><img src="https://avatars.githubusercontent.com/u/43450242?v=4" width="20"></a>                                       | Laniakea1999    | 4             | [**20m**](https://app.flowwer.dev/charts/review-time/~(u~(i~'43450242~n~'Laniakea1999)~p~180~r~(~(d~'qq7oa5~t~'1n)~(d~'qpupsq~t~'d4)~(d~'qqeyk1~t~'2rx7)~(d~'qrmbac~t~'1h5))))                                                                                                                                                                                                                                                                                                                                                                       | 1              |
## Pull reviewers stats
Stats for the last 180 days:
|                                                                                                                                                                 | User            | Total reviews | Median time to review                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       | Total comments |
| --------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------- | ------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------- |
| <a href="https://github.com/Huntsekkern"><img src="https://avatars.githubusercontent.com/u/44240415?v=4" width="20"></a>                                        | Huntsekkern     | 13            | [50m](https://app.flowwer.dev/charts/review-time/~(u~(i~'44240415~n~'Huntsekkern)~p~180~r~(~(d~'qpumtn~t~'1745)~(d~'qpt0zy~t~'2wmx)~(d~'qpt88n~t~'33vm)~(d~'qptd1d~t~'29r)~(d~'qptd3y~t~'2cc)~(d~'qpt4tp~t~'2b1)~(d~'qpuo7b~t~'mk)~(d~'qq6uip~t~'3j5)~(d~'qq8faa~t~'qv)~(d~'qq8fiz~t~'bv)~(d~'qqjij5~t~'i6m)~(d~'qrihqg~t~'1wf3)~(d~'qrm9wa~t~'17x)~(d~'qrma4k~t~'1g7)~(d~'qrmajv~t~'1vi)~(d~'qrni3f~t~'15)~(d~'qqkomy~t~'z8i)~(d~'qqkozj~t~'zl3)~(d~'qqkp4h~t~'zq1)~(d~'qqvo8r~t~'14d)~(d~'qrmb5j~t~'1cc)~(d~'qrs0pv~t~'3yyp))))                                                           | **31**         |
| <a href="https://github.com/so7fie"><img src="https://avatars.githubusercontent.com/u/10992808?v=4" width="20"></a>                                             | so7fie          | 13            | [19h 10m](https://app.flowwer.dev/charts/review-time/~(u~(i~'10992808~n~'so7fie)~p~180~r~(~(d~'qpuy0b~t~'1v)~(d~'qq0p5z~t~'5hby)~(d~'qq0t0v~t~'5m7q)~(d~'qq0t2c~t~'5m97)~(d~'qqew7j~t~'2pkp)~(d~'qqf92s~t~'32fy)~(d~'qqjbho~t~'1ux)~(d~'qqy5l4~t~'5im1)~(d~'qq7qp0~t~'hf)~(d~'qrm6v4~t~'6ai)~(d~'qqkh75~t~'dr0)~(d~'qqrud7~t~'7qx2)~(d~'qr8kh3~t~'1kfr)~(d~'qrm9jc~t~'1npk)~(d~'qrmbez~t~'1ls)~(d~'qrpf7u~t~'11ut)~(d~'qrr5wv~t~'1e1c)~(d~'qrnshx~t~'193))))                                                                                                                                | 24             |
| <a href="https://github.com/RaulinN"><img src="https://avatars.githubusercontent.com/u/44499915?u=c029696ff2de0902f905fbf35188359e94e08d88&v=4" width="20"></a> | RaulinN         | **15**        | [3h 48m](https://app.flowwer.dev/charts/review-time/~(u~(i~'44499915~n~'RaulinN)~p~180~r~(~(d~'qptcqf~t~'1yt)~(d~'qpq5rb~t~'5xjz)~(d~'qq73ig~t~'jy6)~(d~'qq7vmj~t~'1t5)~(d~'qpuwak~t~'8pt)~(d~'qq73am~t~'28)~(d~'qqhz4w~t~'87bj)~(d~'qq73x3~t~'8mkh)~(d~'qq821x~t~'lj5)~(d~'qqr6wy~t~'aju)~(d~'qreyev~t~'f)~(d~'qrni37~t~'x)~(d~'qq7qbn~t~'1bz)~(d~'qqj0lo~t~'u5)~(d~'qqjmtu~t~'n2b)~(d~'qrm5d5~t~'2qe)~(d~'qrrzpp~t~'3mco)~(d~'qrsjsf~t~'7tp)~(d~'qrryuq~t~'428q)~(d~'qrrz93~t~'42n3)~(d~'qrs0cu~t~'3ylo))))                                                                               | 24             |
| <a href="https://github.com/Snowflake8"><img src="https://avatars.githubusercontent.com/u/44499990?v=4" width="20"></a>                                         | Snowflake8      | 14            | [3h 8m](https://app.flowwer.dev/charts/review-time/~(u~(i~'44499990~n~'Snowflake8)~p~180~r~(~(d~'qpt3s9~t~'2zf8)~(d~'qpx29n~t~'1vgi)~(d~'qq0q7g~t~'5jeb)~(d~'qq6k9y~t~'8hk)~(d~'qq6kh9~t~'8ov)~(d~'qq6r5v~t~'6b)~(d~'qqdm38~t~'1fge)~(d~'qq6ncv~t~'8609)~(d~'qq8el9~t~'y2h)~(d~'qqjh1g~t~'gox)~(d~'qqjh87~t~'7lg)~(d~'qqjlvf~t~'b)~(d~'qqsnpq~t~'qn)~(d~'qqxzb3~t~'ikt)~(d~'qrm4bb~t~'1ljt)~(d~'qrnhv0~t~'7)~(d~'qrkbo5~t~'88t)~(d~'qrm4eh~t~'1rq)~(d~'qro2s6~t~'666))))                                                                                                                    | 20             |
| <a href="https://github.com/DGoedtkindtEPFL"><img src="https://avatars.githubusercontent.com/u/61052800?v=4" width="20"></a>                                    | DGoedtkindtEPFL | 14            | [19h 49m](https://app.flowwer.dev/charts/review-time/~(u~(i~'61052800~n~'DGoedtkindtEPFL)~p~180~r~(~(d~'qpuy16~t~'2q)~(d~'qpuoy1~t~'co)~(d~'qpup0b~t~'19at)~(d~'qplma9~t~'1e2x)~(d~'qpuugt~t~'am9h)~(d~'qq6s0w~t~'8gm)~(d~'qq6rfm~t~'fn8)~(d~'qqjjxz~t~'9s4m)~(d~'qq9yj2~t~'1i5c)~(d~'qqjmgc~t~'dy)~(d~'qqrtyv~t~'xlr)~(d~'qqxoq1~t~'21ln)~(d~'qqioj1~t~'2x8a)~(d~'qqionj~t~'2xcs)~(d~'qqioou~t~'2xe3)~(d~'qqipcq~t~'2y1z)~(d~'qqru2n~t~'7qmi)~(d~'qrm4pz~t~'1iw7)~(d~'qrm4za~t~'1j5i)~(d~'qrm5eq~t~'1jky)~(d~'qrmfv4~t~'1u1c)~(d~'qrmhyq~t~'1cp)~(d~'qrnxgy~t~'1wd)~(d~'qrsrti~t~'4q2c)))) | 8              |
| <a href="https://github.com/Laniakea1999"><img src="https://avatars.githubusercontent.com/u/43450242?v=4" width="20"></a>                                       | Laniakea1999    | 4             | [**20m**](https://app.flowwer.dev/charts/review-time/~(u~(i~'43450242~n~'Laniakea1999)~p~180~r~(~(d~'qpupsq~t~'d4)~(d~'qq7oa5~t~'1n)~(d~'qqeyk1~t~'2rx7)~(d~'qrmbac~t~'1h5))))                                                                                                                                                                                                                                                                                                                                                                                                              | 1              |

### CODE REVIEW

**Comment by Huntsekkern** [comment]
I did not properly review the files, but the diff-coverage is pretty low, do you think it's something we can fix? Or we just to accept it for this PR?

**Comment by so7fie** [comment]
> I did not properly review the files, but the diff-coverage is pretty low, do you think it's something we can fix? Or we just to accept it for this PR?

Yes this is definitely something we can fix, I am currently writing tests for this PR. But still, this PR needs to be merged quickly because it appears that making general Query implementing Serializable is not something we can do (ask Dario for details). Thus the main branch is wrong and we should correct it before some people begin to work with Queries on their own branch.

**Review by DGoedtkindtEPFL — APPROVED** [review]

**Comment by so7fie** [comment]
Pourquoi les testes ne passent plus d'un coup??

**Comment by RaulinN** [comment]
C'est un problème assez commun. Tout fonctionne bien sur ta branche, et lorsque tu merge la branche main sur la tienne (ce que tu viens de faire), les changements faits par les autres membres rendent ton code "faux". Je pense que le changement ne devrait pas être trop long, il ne s'agit que d'un null poitner expection

File: app/src/main/java/com/github/polybooks/core/database/interfaces/BookDatabase.kt
Comment (RaulinN) [inline app/src/main/java/com/github/polybooks/core/database/interfaces/BookDatabase.kt:90]: Je ne suis pas sûr de comprendre ce concept de "settings", est-ce que quelqu'un pourrait m'expliquer ce que c'est/pourquoi c'est utile?

File: app/src/main/java/com/github/polybooks/core/database/interfaces/BookDatabase.kt
Comment (RaulinN) [inline app/src/main/java/com/github/polybooks/core/database/interfaces/BookDatabase.kt:87]: Probably change the name "isbn13" to "isbn" to match what we have in the latests two PR : #108 and #109 (they changed isbn13 to isbn)

```suggestion
        val isbn: String?,
```

**Review by RaulinN — COMMENTED** [review]
Wasn't able to review the code thoroughly because I do not understand the concept of Settings as explained in one of my messages. Will do a review at a later date tho

**Review by Huntsekkern — COMMENTED** [review]
As Nicolas, I'm not 100% sure of the reasons behind those changes, so it might be good to document it in the code.
But to try to answer Nicolas, I **think** it has to do with the interaction between Firebase and Serializable. Something like them not working well together, so while Query stays related to Firebase, Settings are decorellated from Firebase and as such can be Serializable? Sophie should confirm this though, don't take my word for granted.

Other than that, I see no issues in the code, it just needs to pass the tests and then ensure that the test-coverage is alright.

File: app/src/main/java/com/github/polybooks/core/database/interfaces/BookDatabase.kt
Comment (DGoedtkindtEPFL) [reply app/src/main/java/com/github/polybooks/core/database/interfaces/BookDatabase.kt:90]: Un Setting (peut-être que le choix de mot n'est pas le meilleur) est un objet sérialisable qui permet d'extraire les Settings d'une query, le transmettre avec un intent, et reconstruire une query semblable dans une autre activité.

**Comment by so7fie** [comment]
> As Nicolas, I'm not 100% sure of the reasons behind those changes, so it might be good to document it in the code.
> But to try to answer Nicolas, I **think** it has to do with the interaction between Firebase and Serializable. Something like them not working well together, so while Query stays related to Firebase, Settings are decorellated from Firebase and as such can be Serializable? Sophie should confirm this though, don't take my word for granted.
> 
> Other than that, I see no issues in the code, it just needs to pass the tests and then ensure that the test-coverage is alright.

Yes exactly, I will add some more comments in the code to explicit that. The tests that do not pass are because Joshua changed the `Sale` in firebase, so I will wait until his pr is merged into main, then I will pull from main and hopefully everything will be working.

**Review by Huntsekkern — APPROVED** [review]
Ok for me!

**Pull request merged**
